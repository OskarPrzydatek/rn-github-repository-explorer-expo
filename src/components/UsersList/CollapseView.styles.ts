import { StyleSheet } from 'react-native';

export const collapseStyles = StyleSheet.create({
  collapse: {
    overflow: 'hidden',
  },
  contentWrapper: {
    width: '100%',
    position: 'absolute',
  },
});
