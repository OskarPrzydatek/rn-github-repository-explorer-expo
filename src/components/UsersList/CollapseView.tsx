import React, { ReactNode, useState } from 'react';
import { LayoutChangeEvent, View } from 'react-native';
import Animated, {
  useAnimatedStyle,
  useSharedValue,
  withTiming,
} from 'react-native-reanimated';
import { collapseStyles } from '@/components/UsersList/CollapseView.styles';

interface ICollapseView {
  expanded: boolean;
  children: ReactNode;
}

export const CollapseView = ({ expanded, children }: ICollapseView) => {
  const animatedHeight = useSharedValue(0);
  const [height, setHeight] = useState(0);

  const onLayout = (event: LayoutChangeEvent) => {
    const layoutHeight = event.nativeEvent.layout.height;

    if (layoutHeight > 0 && height !== layoutHeight) {
      setHeight(layoutHeight);
    }
  };

  const collapsableAnimatedStyle = useAnimatedStyle(() => {
    animatedHeight.value = expanded ? withTiming(height) : withTiming(0);

    return {
      height: animatedHeight.value,
    };
  }, [expanded, height]);

  return (
    <Animated.View style={[collapsableAnimatedStyle, collapseStyles.collapse]}>
      <View style={collapseStyles.contentWrapper} onLayout={onLayout}>
        {children}
      </View>
    </Animated.View>
  );
};
