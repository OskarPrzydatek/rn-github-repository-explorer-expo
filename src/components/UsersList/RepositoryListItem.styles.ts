import { colors } from '@/constants';
import { StyleSheet } from 'react-native';

export const repositoryListItemStyles = StyleSheet.create({
  repositoryListItem: {
    paddingTop: 8,
    paddingRight: 8,
    paddingBottom: 32,
    paddingLeft: 24,
    gap: 8,
  },
  row: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  name: {
    width: '80%',
  },
  stargazersCounter: {
    flexDirection: 'row',
    gap: 12,
    alignItems: 'center',
  },
  stargazersIcon: {
    width: 16,
    height: 16,
  },
  divider: {
    borderBottomWidth: 1,
    borderBottomColor: colors.darkGrey,
  },
  description: {
    width: '85%',
  },
});
