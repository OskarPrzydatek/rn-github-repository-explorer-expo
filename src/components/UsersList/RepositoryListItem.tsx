import React from 'react';
import { Text, View } from 'react-native';
import { repositoryListItemStyles } from '@/components/UsersList/RepositoryListItem.styles';
import { Image } from 'expo-image';

interface IRepositoryListItem {
  name: string;
  description: string | null;
  stargazersCount: number;
  isLastItem: boolean;
}

const STAR_ICON = require('../../../assets/svg/star.svg');

export const RepositoryListItem = ({
  name,
  description,
  stargazersCount,
  isLastItem,
}: IRepositoryListItem) => {
  const isItemDividerStyles = !isLastItem && repositoryListItemStyles.divider;

  return (
    <View
      style={[repositoryListItemStyles.repositoryListItem, isItemDividerStyles]}
    >
      <View style={repositoryListItemStyles.row}>
        <Text numberOfLines={1} style={repositoryListItemStyles.name}>
          {name}
        </Text>

        <View style={repositoryListItemStyles.stargazersCounter}>
          <Text>{stargazersCount}</Text>

          <Image
            style={repositoryListItemStyles.stargazersIcon}
            source={STAR_ICON}
          />
        </View>
      </View>

      <Text numberOfLines={3} style={repositoryListItemStyles.description}>
        {description ?? 'No repository description'}
      </Text>
    </View>
  );
};
