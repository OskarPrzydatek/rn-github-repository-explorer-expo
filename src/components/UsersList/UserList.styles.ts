import { StyleSheet } from 'react-native';

export const userListStyles = StyleSheet.create({
  contentContainerStyle: {
    gap: 8,
    paddingBottom: 64,
  },
});
