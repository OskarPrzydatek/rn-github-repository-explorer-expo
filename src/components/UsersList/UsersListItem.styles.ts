import { colors } from '@/constants';
import { StyleSheet } from 'react-native';

export const userListItemStyles = StyleSheet.create({
  userListItem: {
    borderWidth: 1,
    borderRadius: 6,
    borderColor: colors.darkGrey,
  },
  collapseButton: {
    padding: 12,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    backgroundColor: colors.grey,
  },
  collapseButtonClose: {
    borderRadius: 6,
  },
  collapseButtonOpen: {
    borderTopRightRadius: 6,
    borderTopLeftRadius: 6,
    borderBottomWidth: 1,
    borderBottomColor: colors.darkGrey,
  },
  collapseButtonChevron: {
    width: 10,
    height: 10,
  },
  collapseButtonChevronOpen: {
    transform: [
      {
        rotate: '180deg',
      },
    ],
  },
});
