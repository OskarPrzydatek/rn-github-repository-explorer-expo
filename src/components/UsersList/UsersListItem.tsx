import React from 'react';

import {
  ActivityIndicator,
  FlatList,
  ListRenderItem,
  Pressable,
  Text,
  View,
} from 'react-native';
import { Image } from 'expo-image';

import { IRepository } from '@/api/models';
import { RepositoryListItem } from '@/components/UsersList/RepositoryListItem';
import { userListItemStyles } from '@/components/UsersList/UsersListItem.styles';
import { colors } from '@/constants';

interface IUsersListItem {
  login: string;
  userRepositories: IRepository[] | undefined;
  isUserRepositoriesListExpanded: boolean;
  isLoadingRepositoriesByUserData: boolean;
  fetchRepositoriesByUserError: Error | null;
  onPressExpand: () => void;

  expandingPressableTestID?: string;
  loadingViewTestID?: string;
  repositoriesListTestID?: string;
  noRepositoriesFoundViewTestID?: string;
  errorViewTestID?: string;
}

const CHEVRON_ICON = require('../../../assets/svg/chevron.svg');

export const UsersListItem = ({
  isUserRepositoriesListExpanded,
  login,
  userRepositories,
  isLoadingRepositoriesByUserData,
  fetchRepositoriesByUserError,
  onPressExpand,

  expandingPressableTestID,
  loadingViewTestID,
  repositoriesListTestID,
  noRepositoriesFoundViewTestID,
  errorViewTestID,
}: IUsersListItem) => {
  const expandStateButtonStyle = isUserRepositoriesListExpanded
    ? userListItemStyles.collapseButtonOpen
    : userListItemStyles.collapseButtonClose;
  const expandStateChevronRotateStyle =
    isUserRepositoriesListExpanded &&
    userListItemStyles.collapseButtonChevronOpen;

  const isExpandedWithLoading =
    isUserRepositoriesListExpanded && isLoadingRepositoriesByUserData;
  const isExpandedWithRepositories =
    isUserRepositoriesListExpanded &&
    userRepositories &&
    userRepositories?.length > 0;
  const isExpandedWithNoRepositories =
    isUserRepositoriesListExpanded &&
    userRepositories &&
    userRepositories?.length === 0;
  const isExpandedWithError =
    isUserRepositoriesListExpanded && fetchRepositoriesByUserError;

  const renderItemForRepositoriesList: ListRenderItem<IRepository> = ({
    item,
    index,
  }) => {
    const isLastItem = index + 1 === userRepositories?.length;

    return (
      <RepositoryListItem
        name={item.name}
        description={item.description}
        stargazersCount={item.stargazers_count}
        isLastItem={isLastItem}
      />
    );
  };

  return (
    <View style={userListItemStyles.userListItem}>
      <Pressable
        style={[userListItemStyles.collapseButton, expandStateButtonStyle]}
        onPress={onPressExpand}
        testID={expandingPressableTestID}
      >
        <Text>{login}</Text>

        <Image
          style={[
            userListItemStyles.collapseButtonChevron,
            expandStateChevronRotateStyle,
          ]}
          source={CHEVRON_ICON}
        />
      </Pressable>

      {isExpandedWithLoading && (
        <View
          style={userListItemStyles.isLoadingRepositories}
          testID={loadingViewTestID}
        >
          <ActivityIndicator color={colors.black} />
        </View>
      )}

      {isExpandedWithRepositories && (
        <FlatList
          data={userRepositories}
          renderItem={renderItemForRepositoriesList}
          scrollEnabled={false}
          testID={repositoriesListTestID}
        />
      )}

      {isExpandedWithNoRepositories && (
        <View
          style={userListItemStyles.noRepositoriesFound}
          testID={noRepositoriesFoundViewTestID}
        >
          <Text>No repositories found</Text>
        </View>
      )}

      {isExpandedWithError && (
        <View style={userListItemStyles.fetchError} testID={errorViewTestID}>
          <Text>Oops! Somethiong went wrong :(</Text>
        </View>
      )}
    </View>
  );
};
