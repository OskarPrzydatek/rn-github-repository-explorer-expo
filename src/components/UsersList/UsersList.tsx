import React from 'react';

import { FlatList, ListRenderItem } from 'react-native';

import { UsersListItem } from '@/components/UsersList/UsersListItem';
import { userListStyles } from '@/components/UsersList/UsersList.styles';
import { IRepository, IUser } from '@/api/models';

interface IUsersList {
  users: IUser[] | undefined;
  userRepositories: IRepository[] | undefined;
  fetchRepositoriesByUserError: Error | null;
  isLoadingRepositoriesByUserData: boolean;
  userRepositoriesListExpanded: number | null;
  onPressUserRepositoriesListExpand: (username: string, index: number) => void;

  testID?: string;
  usersListItemExpandingPressableTestID?: string;
}

export const UsersList = ({
  users,
  userRepositories,
  isLoadingRepositoriesByUserData,
  fetchRepositoriesByUserError,
  userRepositoriesListExpanded,
  onPressUserRepositoriesListExpand,

  testID,
  usersListItemExpandingPressableTestID,
}: IUsersList) => {
  const renderItemForUsersList: ListRenderItem<IUser> = ({ item, index }) => {
    const isUserRepositoriesListExpanded =
      userRepositoriesListExpanded === index;

    const handleOnPressUserRepositoriesListExpand = () =>
      onPressUserRepositoriesListExpand(item.login, index);

    return (
      <UsersListItem
        login={item.login}
        userRepositories={userRepositories}
        isUserRepositoriesListExpanded={isUserRepositoriesListExpanded}
        isLoadingRepositoriesByUserData={isLoadingRepositoriesByUserData}
        fetchRepositoriesByUserError={fetchRepositoriesByUserError}
        onPressExpand={handleOnPressUserRepositoriesListExpand}
        expandingPressableTestID={`${index}-${usersListItemExpandingPressableTestID}`}
      />
    );
  };

  return (
    <FlatList
      data={users}
      renderItem={renderItemForUsersList}
      contentContainerStyle={userListStyles.contentContainerStyle}
      showsVerticalScrollIndicator={false}
      testID={testID}
    />
  );
};
