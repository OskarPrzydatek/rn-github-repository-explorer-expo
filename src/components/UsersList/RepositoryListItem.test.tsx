import { render } from '@testing-library/react-native';

import { RepositoryListItem } from '@/components/UsersList/RepositoryListItem';

jest.mock('expo-image', () => {
  const actualExpoImage = jest.requireActual('expo-image');
  const { Image } = jest.requireActual('react-native');

  return { ...actualExpoImage, Image };
});

describe('RepositoryListItem', () => {
  it('should match snapshot', () => {
    const tree = render(
      <RepositoryListItem
        name={''}
        description={null}
        stargazersCount={0}
        isLastItem={false}
      />,
    );

    expect(tree).toMatchSnapshot();
  });
});
