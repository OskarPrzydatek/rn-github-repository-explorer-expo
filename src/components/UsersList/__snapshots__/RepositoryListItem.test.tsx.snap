// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepositoryListItem should match snapshot 1`] = `
<View
  style={
    [
      {
        "gap": 8,
        "paddingBottom": 32,
        "paddingLeft": 24,
        "paddingRight": 8,
        "paddingTop": 8,
      },
      {
        "borderBottomColor": "#d0d7de",
        "borderBottomWidth": 1,
      },
    ]
  }
>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      numberOfLines={1}
      style={
        {
          "width": "80%",
        }
      }
    />
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "gap": 12,
        }
      }
    >
      <Text>
        0
      </Text>
      <Image
        source={1}
        style={
          {
            "height": 16,
            "width": 16,
          }
        }
      />
    </View>
  </View>
  <Text
    numberOfLines={3}
    style={
      {
        "width": "85%",
      }
    }
  >
    No repository description
  </Text>
</View>
`;
